### DISCUSSION TITLE
Should all major types be interfaces?

### DISCUSSION THREAD

**Discussion by madskristensen** [VsixCommunity/Community.VisualStudio.Toolkit#85] [Ideas]
Types such as `VS.MessageBox` and `VS.Solution` return classes, but should they return an interface instead? That would make mocking and testing easier. It adds more weight and files to the toolkit, so do you think it's worth doing?

**Comment by CZEMacLeod** [comment]
I guess it depends on whether people need to do mocking and testing of these methods. My experience is that unless you have different implementation classes, or a fixed/simple interface that *really really* isn't going to change, interfaces just add a lot of work to maintain for very little gain.

**Reply by madskristensen** [reply]
That's my experience as well and why I didn't originally want to use them. I'm just wondering if there are other benefits aside from mocking/testing.

**Comment by yannduran** [comment]
@madskristensen I've already noted that I have changes to the messagebox code coming as soon as I can as health permits.

**Comment by reduckted** [comment]
I think it's absolutely worth doing. In fact, I would say we _need_ to do it if we want to make testing of extensions easier. I've been playing with some ideas for testing helpers and I hope to have a prototype up and running in the next few days. One of the key parts to the whole idea is having the `VS` class return interfaces instead of classes, so that we can provide implementations for tests.
